design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,324 @@
+When and if a feature is targeted for deprecation and retirement,
+it needs to pass a deprecation process, depending on its current
+release stage (Alpha, Beta, GA).
+
+For more details, see [here](#deprecation-and-removal).
+
+#### Release Stage Transition Table
+The following table summarized the different release stages with their
+transition requirements and restrictions.
+
+| Stage          | Period range    | F.Gate | Removal Availability       |
+|----------------|-----------------|--------|----------------------------|
+| Alpha          | 1 to 2 releases | YES    | Between **minor** releases |
+| Beta           | 1 to 3 releases | YES    | Between **minor** releases |
+| GA             | -               | NO     | Between **major** releases |
+
+Through Alpha and Beta feature releases, a FG must be set in order
+for the feature to function.
+By default, no FG is specified, therefore the feature is by disabled.
+
+If a feature is not able to transition to the next stage in the defined period,
+it should be removed automatically.
+
+> **Note**: Exceptions to the period range may apply
+> if 2/3 of active maintainers come to agreement to prolong
+> a specific feature.
+
+### Deprecation and Removal
+One reason for features to go through the Alpha and Beta stages,
+is the opportunity to examine their usefulness and adoption.
+Same goes with major releases that intentionally allow breaking
+backward compatibility (as specified by [semver](https://semver.org/)).
+
+Therefore, it is only natural that some features will not graduate
+between the stages, or will be found irrelevant after some time and be
+removed when transitioning between major releases.
+
+#### Major Releases
+KubeVirt follows semver versioning, in which major versions may
+break API compatibility. Therefore, discontinuation of features
+is somehow simpler when incrementing the major version.
+
+However, this is not without a cost.
+When a new major release is introduced, the previous one is still maintained
+and supported, something that does not exist with minor releases.
+
+#### The Deprecation Flow (for Minor releases)
+Only Alpha and Beta features can be removed during a minor release.
+
+These are the steps needed to deprecate & remove a feature:
+- Proposal: Prepare a proposal to remove a feature with proper
+  reasoning, phases, exact timelines and functional alternatives (if any).
+  The proposal should be reviewed and approved.
+- Notification: Notify the project community about the feature
+  discontinuation based on the approved proposal.
+  All details of the plan should be provided to allow users and possibly
+  down-stream projects to adjust.
+  Use all community media options to distribute this information
+  (e.g. mailing list, slack channel, community meetings).
+- Deprecation warnings: Add deprecation warnings at runtime to warn users
+  that the feature is planned to be removed.
+  Warnings should be raised when:
+  - Feature API fields are accessed.
```

Comment:
Can we technically guarantee that?
For example, what happens if someone reads a field? how can we trigger a warning in that case?

Reply:
I think the meaning is when used on definition.
If someone tries to read them, it is after it was written to.

Reply:
is there work needed to achieve this proposal? if so do we have an issue for this?

Reply:
WDYT about
```suggestion
  - Feature API fields are being used in a non-default way.
```

Reply:
What do we need to resolve this conversation?

Reply:
I prefer to leave the wording as is.
As @iholder101 lgtm this proposal, I suspect he does not insist on it, therefore I will resolve it.